File: P2_Complejidad/busq_lineal.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 1000000

int busquedaLineal(int a[], int n, int x){
    for(int i = 0; i < n; i++){
        if(a[i] == x)
            return i;
    }
    return -1;
}

void llenarOrdenado(int a[]){
    for(int i = 0; i < N; i++)
        a[i] = i + 1;
}

void llenarDesordenado(int a[]){
    for(int i = 0; i < N; i++)
        a[i] = i + 1;

    for(int i = 0; i < N; i++){
        int j = rand() % N;
        int tmp = a[i];
        a[i] = a[j];
        a[j] = tmp;
    }
}

int main(){
    int opcion, x;
    int *a = (int*)malloc(N * sizeof(int));
    clock_t ini, fin;
    double tiempo;
    int pos;
    int memoria = N * sizeof(int);

    srand(time(NULL));

    do{
        printf("\nBUSQUEDA LINEAL DEL 1 AL 1,000,000\n");
        printf("1. Busqueda en lista ordenada\n");
        printf("2. Busqueda en lista desordenada\n");
        printf("3. Salir\n");
        printf("Opcion: ");
        scanf("%d", &opcion);

        if(opcion == 3) break;

        printf("Numero a buscar: ");
        scanf("%d", &x);

        system("cls"); // Limpia pantalla

        if(opcion == 1){
            llenarOrdenado(a);
            printf("LISTA ORDENADA\n");
        }
        else if(opcion == 2){
            llenarDesordenado(a);
            printf("LISTA DESORDENADA\n");
        }
        else{
            printf("Opcion invalida\n");
            continue;
        }

        ini = clock();
        pos = busquedaLineal(a, N, x);
        fin = clock();

        tiempo = (double)(fin - ini) / CLOCKS_PER_SEC;

        if(pos != -1)
            printf("Elemento %d encontrado en la posicion %d del arreglo\n", x, pos);
        else
            printf("Elemento no encontrado\n");

        printf("Tiempo de ejecucion: %f segundos\n", tiempo);
        printf("Memoria usada: %d bytes\n", memoria);

        printf("\nTabla comparativa:\n");
        printf("Tamanio\tTiempo(s)\tMemoria(bytes)\n");
        printf("%d\t%f\t%d\n", N, tiempo, memoria);

        system("pause");

    }while(1);

    free(a);
    return 0;
}

File: P2_Complejidad/busq_lineal.py
import time
import random
import os
import sys

N = 1_000_000

def busqueda_lineal(lista, x):
    for i in range(len(lista)):
        if lista[i] == x:
            return i
    return -1

def llenar_ordenado():
    return list(range(1, N + 1))

def llenar_desordenado():
    lista = list(range(1, N + 1))
    random.shuffle(lista)
    return lista

def memoria_usada():
    # Aproximacion del consumo de memoria
    return sys.getsizeof([0]) * N

while True:
    print("\nBUSQUEDA LINEAL DEL 1 AL 1,000,000")
    print("1. Busqueda en lista ordenada")
    print("2. Busqueda en lista desordenada")
    print("3. Salir")
    
    opcion = input("Opcion: ")

    if opcion == "3":
        break

    x = int(input("Numero a buscar: "))

    os.system("cls" if os.name == "nt" else "clear")

    if opcion == "1":
        lista = llenar_ordenado()
        print("LISTA ORDENADA")
    elif opcion == "2":
        lista = llenar_desordenado()
        print("LISTA DESORDENADA")
    else:
        print("Opcion invalida")
        continue

    inicio = time.perf_counter()
    pos = busqueda_lineal(lista, x)
    fin = time.perf_counter()

    tiempo = fin - inicio
    memoria = memoria_usada()

    if pos != -1:
        print(f"Elemento {x} encontrado en la posicion {pos} de la lista")
    else:
        print("Elemento no encontrado")

    print(f"Tiempo de ejecucion: {tiempo:.6f} segundos")
    print(f"Memoria aproximada usada: {memoria} bytes")

    print("\nTabla comparativa:")
    print("Tamanio\tTiempo(s)\tMemoria(bytes)")
    print(f"{N}\t{tiempo:.6f}\t{memoria}")

    input("\nPresiona Enter para continuar...")